scripts/setup_haskell.hs

```diff
@@ -0,0 +1,355 @@
+
+data HvnArgs = HvnArgs
+  { hvnArgsNoHoogleDb       :: Bool
+  , hvnArgsNoHelperBinaries :: Bool
+  } deriving (Show)
+
+main :: IO ()
+main = do
+  HvnArgs{..} <- Turtle.options "Haskell Vim Now - setup Haskell specifics"
+                    cliParser
+  print HvnArgs{..}
+  hvnCfgHome <- hvnHomeDir
+  let hvnCfgDest = hvnCfgHome </> "haskell-vim-now"
+      hvnCfgHoogleDb = not hvnArgsNoHoogleDb
+      hvnCfgHelperBinaries = not hvnArgsNoHelperBinaries
+  runReaderT setup HvnConfig{..}
+  Turtle.exit Turtle.ExitSuccess
+
+cliParser :: Turtle.Parser HvnArgs
+cliParser = HvnArgs
+        <$> Turtle.switch "no-hoogle" 'g'
+              "Disable Hoogle database generation."
+        <*> Turtle.switch "no-helper-bins" 'b'
+              "Disable install of helper binaries (mainly for CI)."
+
+hvnHomeDir :: (MonadIO m) => m FilePath
+hvnHomeDir = do
+  mXdgConfigHome <- Turtle.need "XDG_CONFIG_HOME"
+  maybe
+    (fmap (</> ".config") Turtle.home)
+    (pure . textToFilePath)
+    mXdgConfigHome
+
+setup :: (MonadIO m, MonadReader HvnConfig m) => m ()
+setup = do
+  setupHaskell
+  msg "HASKELL VIM NOW install of Haskell specifics successfully finished"
+
+setupHaskell :: (MonadIO m, MonadReader HvnConfig m) => m ()
+setupHaskell = do
+  HvnConfig{..} <- ask
+  hasStack <- hasExecutable "stack"
+  unless hasStack $ do
+    err "Installer requires Stack - installation instructions:"
+    msg "  http://docs.haskellstack.org/en/stable/README/#how-to-install"
+  when hasStack $ do
+    msg "Setting up GHC if needed..."
+    stackSetupResult <- Turtle.shell "stack setup --verbosity warning" empty
+    case stackSetupResult of
+      (Turtle.ExitFailure retCode) -> do
+        err $ "Stack setup failed with error " <> (Text.pack . show $ retCode)
+        Turtle.exit (Turtle.ExitFailure 1)
+      Turtle.ExitSuccess -> do
+        stackBinPath <-
+          commandSubstitution "stack --verbosity 0 path --local-bin"
+        stackGlobalDir <-
+          commandSubstitution "stack --verbosity 0 path --stack-root"
+        stackGlobalConfig <-
+          commandSubstitution "stack --verbosity 0 path --config-location"
+        stackResolver <- stackResolverText . textToFilePath $ stackGlobalConfig
+        detail $ "Stack bin path: " <> stackBinPath
+        detail $ "Stack global path: " <> stackGlobalDir
+        detail $ "Stack global config location: " <> stackGlobalConfig
+        detail $ "Stack resolver: " <> stackResolver
+        let emptyStackPath = any Text.null
+                               [ stackBinPath
+                               , stackGlobalDir
+                               , stackGlobalConfig
+                               ]
+        when emptyStackPath $ do
+          err "Incorrect stack paths."
+          Turtle.exit (Turtle.ExitFailure 1)
+        let stackBinUnderCfgDest = hvnCfgDest </> ".stack-bin"
+        mkDirLink (textToFilePath stackBinPath) stackBinUnderCfgDest
+        when hvnCfgHelperBinaries $ do
```

Comment:
This seemed a little magical at first glance. It was only after looking back up and seeing `HvnConfig{..} <- ask` where I realized that a few of these local labels are getting filled out at once.

Just for easier comprehension what if you did `cfg <- ask` at the top and then in places like this reference `hvnCfgHelperBinaries cfg`? I know your knowledge of Haskell probably makes this seem like a kludge, but I'm just thinking of other people trying to read the code.

Feel free to push back on my recommendations as well because I can tell you have good Haskell style.